aea/registries/base.py

```diff
@@ -225,292 +237,165 @@ def _add_protocol(self, protocol_directory: Path):
+        if len(name_to_handlers) == 0:
+            self._items.pop(skill_id, None)
 
-    def unregister(self, skill_id: SkillId) -> None:
-        """
-        Unregister a behaviour.
-
-        :param skill_id: the skill id.
-        :return: None
+    def fetch(self, item_id: Tuple[SkillId, str]) -> Optional[SkillComponentType]:
         """
-        self._behaviours.pop(skill_id, None)
+        Return a item.
 
-    def fetch(self, skill_id: SkillId) -> Optional[List[Behaviour]]:
+        :return: the list of items
         """
-        Return a behaviour.
+        skill_id = item_id[0]
+        item_name = item_id[1]
+        return self._items.get(skill_id, {}).get(item_name, None)
 
-        :return: the list of behaviours
-        """
-        return self._behaviours.get(skill_id, None)
+    def fetch_by_skill(self, skill_id: SkillId) -> List[Item]:
+        """Fetch all the items of a given skill."""
+        return list(*self._items.get(skill_id, {}).values())
 
-    def fetch_all(self) -> List[Behaviour]:
-        """Fetch all the behaviours."""
+    def fetch_all(self) -> List[SkillComponentType]:
+        """Fetch all the items."""
         return [
-            b
-            for skill_behaviours in self._behaviours.values()
-            for b in skill_behaviours
+            item for skill_id, items in self._items.items() for item in items.values()
         ]
 
+    def unregister_by_skill(self, skill_id: SkillId) -> None:
+        """Unregister all the components by skill."""
+        self._items.pop(skill_id, None)
+
     def setup(self) -> None:
         """
-        Set up the behaviours in the registry.
+        Set up the items in the registry.
 
         :return: None
         """
-        for behaviours in self._behaviours.values():
-            for behaviour in behaviours:
-                behaviour.setup()
+        for item in self.fetch_all():
+            item.setup()
 
     def teardown(self) -> None:
         """
         Teardown the registry.
 
         :return: None
         """
-        for skill_id, behaviours in self._behaviours.items():
-            for behaviour in behaviours:
+        for skill_id, items in self._items.items():
+            for _, item in items.items():
                 try:
-                    behaviour.teardown()
+                    item.teardown()
                 except Exception as e:
                     logger.warning(
-                        "An error occurred while tearing down behaviour {}/{}: {}".format(
-                            skill_id, type(behaviour).__name__, str(e)
+                        "An error occurred while tearing down item {}/{}: {}".format(
+                            skill_id, type(item).__name__, str(e)
                         )
                     )
-        self._behaviours = {}
 
 
-class TaskRegistry(Registry):
-    """This class implements the task registry."""
+class HandlerRegistry(ComponentRegistry[Handler]):
```

Comment:
the `HandlerRegistry` is a special `ComponentRegistry` because it also implements `fetch_by_protocol` and `fetch_by_protocol_and_skill`